autosklearn/ensemble_builder.py

```diff
@@ -30,26 +31,206 @@
+           models to keep.
+           If float, it will be interpreted as the max megabytes allowed of disc space. That
+           is, if the number of ensemble candidates require more disc space than this float
+           value, the worst models will be deleted to keep within this budget.
+           Models and predictions of the worst-performing models will be deleted then.
+           If None, the feature is disabled.
+           It defines an upper bound on the models that can be used in the ensemble.
+        seed: int
+            random seed
+        max_iterations: int
+            maximal number of iterations to run this script
+            (default None --> deactivated)
+        precision: [16,32,64,128]
+            precision of floats to read the predictions
+        memory_limit: Optional[int]
+            memory limit in mb. If ``None``, no memory limit is enforced.
+        read_at_most: int
+            read at most n new prediction files in each iteration
+
+    Returns
+    -------
+        List[Tuple[int, float, float, float]]:
+            A list with the performance history of this ensemble, of the form
+            [[pandas_timestamp, train_performance, val_performance, test_performance], ...]
+    """
+
+    # Create the event as one of the criteria to decide if we move to he next iteration
+    event = dask.distributed.Event(event) if event else None
+
+    # The second criteria is elapsed time
+    elapsed_time = time.time() - start_time
+
+    # The last criteria is the number of iterations
+    iteration = 0
+
+    history = []
+
+    logger = EnsembleBuilder._get_ensemble_logger(backend.temporary_directory)
+
+    while True:
+
+        # First test for termination conditions
+        if time_left_for_ensembles < elapsed_time:
+            logger.info(
+                "Terminate ensemble building as not time is left (run for {}s)".format(
+                    elapsed_time
+                ),
+            )
+            break
+        if event is not None and event.is_set():
+            logger.info(
+                "Terminating ensemble building as SMAC process is done...")
+            break
+        if max_iterations is not None and max_iterations <= iteration:
+            logger.info(
+                "Terminate ensemble building because of max iterations: {} of {}".format(
+                    max_iterations,
+                    iteration
+                )
+            )
+            break
+
+        # Add the result of the run
+        # On the next while iteration, no references to
+        # ensemble builder object, so it should be garbage collected to
+        # save memory while waiting for resources
+        # Also, notice how ensemble nbest is returned, so we don't waste
+        # iterations testing if the deterministic predictions size can
+        # be fitted in memory
+        try:
+            result = EnsembleBuilder(
```

Comment:
If I understand the code correctly, this now has to read all predictions from scratch every time in order to compute the scores?

Reply:
I changed this with a pickled memory as per discussed offline. Please let me know if I misunderstood the problem statement.